README.md
# aml-lib

[![Build Status](https://github.com/cpilatre/aml-lib/actions/workflows/build.yml/badge.svg)](https://github.com/cpilatre/aml-lib/actions?query=workflow%3A%22build%22)
[![Test Status](https://github.com/cpilatre/aml-lib/actions/workflows/test.yml/badge.svg)](https://github.com/cpilatre/aml-lib/actions?query=workflow%3A%22test%22)

## About

aml-lib is a 100% Rust library for dealing with AML (Advanced Mobile Location) messages.

For v1 SMS and HTTPS see [ETSI TS 103 625 V1.1.1 (2019-12)](https://www.etsi.org/deliver/etsi_ts/103600_103699/103625/01.01.01_60/ts_103625v010101p.pdf).

## Main features

- SMS AML v1 and V2 compliance
- Accepts text and data SMS (with binary or Base64 encoded sources)
- HTTPS AML with hmac-sha1 authentification
- Provides a generic AML format

## Installation

Manually add `aml-lib` to your `Cargo.toml` file :

```toml
[dependencies]
aml-lib = "*"
```

Or use [cargo-edit](https://crates.io/crates/cargo-edit) :

```bash
cargo add aml-lib
```

## Usage

- If you receive AML SMS only, you could use `SmsData` structure to parse text sms
 (send by Apple device) or data SMS (send by Android device).

```rust
use aml_lib::SmsData;

let sms_text = String::from(r#"A"ML=1;lt=48.82639;lg=-2.36619;rd=52;top=20191112112928;lc=68;pm=G;si=208201771948415;ei=353472104343540;mcc=208;mnc=20;ml=126"#);
    
let sms_data = SmsData::from_text(&sms_text);
if let Ok(sms) = sms_data {
    println!("{:#?}", sms);
}
```

- If you receive HTTPS AML message only, you could use `HttpsData` structure.

```rust
use aml_lib::HttpsData;

const KEY: &str = "AML";

let https = String::from(r#"v=1&device_number=%2B33611223344&location_latitude=0.85732&location_longitude=-4.26325&location_time=1604912121000&location_accuracy=10.4&location_source=GPS&location_certainty=83&hmac=f64c70eb238bb239e00e8ac8c023bf2b5d3c41dd"#);
if HttpsData::is_authenticated(&https, KEY.as_bytes()) {
    let https_data = HttpsData::from_urlencoded(&https);
    println!("{:#?}", https_data)
} 
```

- For both, SMS and HTTPS, you could use `AmlData` DTO structure to standardize
 AML data.

```rust
use aml_lib::AmlData;

let sms_text = r#"A"ML=1;lt=48.82639;lg=-2.36619;rd=52;top=20191112112928;lc=68;pm=G;si=208201771948415;ei=353472104343540;mcc=208;mnc=20;ml=126"#;
let aml_data_1 = AmlData::from_text_sms(&sms_text);
println!("{:?}", aml_data_1);
```

```rust
use aml_lib::AmlData;

let base64_data_sms = "QVGT2Yvt2PTe7OaiyWK32o597rViMpkLhqPZYjs5uSeD7ehveE8Gi9VgttgMFoPlaLgde9yz4Xbwdu+4m6d7OdzNVqPJZrFdOd2b1XCyWQ5Wy8FoshpNtmuPx71ZDLZru8c9mQ22a7N7MdkM";
let aml_data_2 = AmlData::from_base64_sms(base64_data_sms);
println!("{:?}", aml_data_2);
```

```rust
use aml_lib::{HttpsData, AmlData};

let https = r#"v=1&device_number=%2B33611223344&location_latitude=0.85732&location_longitude=-4.26325&location_time=1604912121000&location_accuracy=10.4&location_source=GPS&location_certainty=83&hmac=f64c70eb238bb239e00e8ac8c023bf2b5d3c41dd"#;
let https_data = HttpsData::from_urlencoded(https);
let aml_data_3: AmlData = https_data.into();
println!("{:?}", aml_data_3);
```

```rust
use aml_lib::{SmsData, AmlData};

let sms_data = SmsData::from_text(sms_text).unwrap();
let aml_data_4 = AmlData::from(sms_data);
println!("{:?}", aml_data_4);
```

## Release History

A short list of features, fixes and changes for each release is available in [CHANGELOG.md](https://github.com/cpilatre/aml-lib/blob/main/CHANGELOG.md).

## Contributing

Anyone is welcome to submit issues and pull requests.

## License

See [LICENSE](LICENSE).


CHANGELOG.md
# Changelog

## 0.1.2 - 2021-06-04

### Added

- `is_validated` property add to SmsData implementation.
- Improvement of documentation.

## 0.1.1 - 2021-06-03

### Added

- Authentification via HMAC-SHA1 added to HTTPS message.

## 0.1.0 - 2021-05-29

✨ First release


tests/verify.rs
use aml_lib::{AmlData, SmsData, HttpsData};

#[test]
fn from_text_sms() {
    let sms_text = String::from(
        r#"A"ML=1;lt=48.82639;lg=-2.36619;rd=52;top=20191112112928;lc=68;pm=G;si=208201771948415;ei=353472104343540;mcc=208;mnc=20;ml=126"#,
    );

    let sms_data = SmsData::from_text(&sms_text);
    if let Ok(sms) = sms_data {
        assert!(sms.latitude == Some(48.82639), "Parsing failed : {:?}", sms);
    } else {
        panic!("Error text SMS");
    }

    let aml = AmlData::from_text_sms(&sms_text).unwrap();
    assert!(
        aml.imei == Some("353472104343540".to_string()),
        "Parsing failed : {:?}",
        aml
    );
}

#[test]
fn sms_is_validate() {
    let sms_text = String::from(
        r#"A"ML=1;lt=48.82639;lg=-2.36619;rd=52;top=20191112112928;lc=68;pm=G;si=208201771948415;ei=353472104343540;mcc=208;mnc=20;ml=126"#,
    );

    let sms_data = SmsData::from_text(&sms_text);
    if let Ok(sms) = sms_data {
        assert!(sms.is_validated, "SMS is not valid : {:?}", sms);
    } else {
        panic!("Error text SMS");
    }
}

#[test]
fn from_text_sms_v2() {
    let sms_text = String::from(
        r#"A"ML=2;en=+15555555555;et=1593187189;lo=-37.42175,-122.08461,2000.1;lt=-9999;lc=68;lz=-100.1,100.1;ls=G;ei=358239059042542;nc=310260;hc=310260;lg=en-US"#,
    );

    let sms_data = SmsData::from_text(&sms_text);
    if let Ok(sms) = sms_data {
        assert!(
            sms.latitude == Some(-37.42175),
            "Parsing failed : {:?}",
            sms
        );
    } else {
        panic!("Error AML SMS v2");
    }
}

#[test]
fn from_data_sms() {
    let input = "415193D98BEDD8F4DEECE6A2C962B7DA8E7DEEB56232990B86A3D9623B39B92783EDE86F784F068BD560B6D80C1683E568B81D7BDCB3E176F076EFB89BA77B39DCCD56A3C966B15D39DD9BD570B2590E56CBC168B21A4DB66B8FC7BD590CB66BBBC73D990DB66BB37B31D90C";
    let decoded = hex::decode(input).expect("Decoding failed");

    let sms_data = SmsData::from_data(&decoded);
    if let Ok(sms) = sms_data {
        assert!(sms.latitude == Some(37.42175), "Parsing failed : {:?}", sms);
    } else {
        panic!("Error data SMS");
    }

    let aml = AmlData::from_data_sms(&decoded).unwrap();
    assert!(
        aml.imei == Some("358239059042542".to_string()),
        "Parsing failed : {:?}",
        aml
    );
}

#[test]
fn from_https() {
    let https = r#"v=1&device_number=%2B447477593102&location_latitude=55.85732&location_longitude=-4.26325&location_time=1476189444435&location_accuracy=10.4&location_source=GPS&location_certainty=83&location_altitude=0.0&location_floor=5&device_model=ABC+ABC+Detente+530&device_imei=354773072099116&device_imsi=234159176307582&device_os=AOS&cell_carrier=&cell_home_mcc=234&cell_home_mnc=15&cell_network_mcc=234&cell_network_mnc=15&cell_id=0213454321"#;

    let aml = AmlData::from_https(&https).unwrap();
    assert!(
        aml.positioning_method == Some("gps".to_string()),
        "Parsing failed : {:?}",
        aml
    );
}

#[test]
fn authenticate() {
    let https = String::from(r#"v=1&device_number=%2B33611223344&location_latitude=0.85732&location_longitude=-4.26325&location_time=1604912121000&location_accuracy=10.4&location_source=GPS&location_certainty=83&hmac=f64c70eb238bb239e00e8ac8c023bf2b5d3c41dd"#);

    assert!(HttpsData::is_authenticated(https, "AML".as_bytes()));
}

src/aml.rs
use crate::{ AmlError, HttpsData, SmsData};
use chrono::{ DateTime, Utc };

#[derive(Debug, Default)]
pub struct AmlData {
    /// See [`SmsData::header`] or [`HttpsData::v`]
    pub version: Option<String>,

    /// See [`SmsData::emergency_number`] or [`HttpsData::emergency_number`]
    pub emergency_number: Option<String>,

    /// See [`HttpsData::source`]
    pub source_of_activation: Option<String>,

    /// See [`SmsData::beginning_of_call`] or [`HttpsData::time`]
    pub beginning_of_call: Option<DateTime<Utc>>,

    /// See [`SmsData::latitude`] or [`HttpsData::location_latitude`]
    pub latitude: Option<f64>,

    /// See [`SmsData::longitude`] or [`HttpsData::location_longitude`]
    pub longitude: Option<f64>,

    /// See [`SmsData::time_of_positioning`] or [`HttpsData::location_time`]
    pub time_of_positioning: Option<DateTime<Utc>>,

    /// See [`SmsData::altitude`] or [`HttpsData::location_altitude`]
    pub altitude: Option<f64>,

    /// See [`HttpsData::location_floor`]
    pub floor: Option<f64>,

    /// See [`SmsData::positioning_method`] or [`HttpsData::location_source`]
    pub positioning_method: Option<String>,

    /// See [`SmsData::accuracy`] or [`HttpsData::location_accuracy`]
    pub accuracy: Option<f64>,

    /// See [`SmsData::vertical_accuracy`] or [`HttpsData::location_vertical_accuracy`]
    pub vertical_accuracy: Option<f64>,

    /// See [`SmsData::level_of_confidence`] or [`HttpsData::location_confidence`]
    pub confidence: Option<f64>,

    /// See [`HttpsData::location_bearing`]
    pub bearing: Option<f64>,

    /// See [`HttpsData::location_speed`]
    pub speed: Option<f64>,

    /// See [`HttpsData::device_number`]
    pub device_number: Option<String>,

    /// See [`HttpsData::device_model`]
    pub model: Option<String>,

    /// See [`SmsData::imsi`] or [`HttpsData::device_imsi`]
    pub imsi: Option<String>,

    /// See [`SmsData::imei`] or [`HttpsData::device_imei`]
    pub imei: Option<String>,

    /// See [`HttpsData::device_iccid`]
    pub iccid: Option<String>,

    /// See [`SmsData::home_mcc`] or [`HttpsData::cell_home_mcc`]
    pub home_mcc: Option<String>,

    /// See [`SmsData::home_mnc`] or [`HttpsData::cell_home_mnc`]
    pub home_mnc: Option<String>,

    /// See [`SmsData::network_mcc`] or [`HttpsData::cell_network_mcc`]
    pub network_mcc: Option<String>,

    /// See [`SmsData::network_mnc`] or [`HttpsData::cell_network_mnc`]
    pub network_mnc: Option<String>,

    /// See [`SmsData::language`] or [`HttpsData::device_languages`]
    pub language: Option<String>,

    /// Where the location comes from: `sms` or `https`
    pub transport: String,
}

impl AmlData {
    /// Parse a HTTPS AML message. See [`HttpsData::from_urlencoded`].
    pub fn from_https(payload: &str) -> Result<Self, AmlError> {
        let https_data = HttpsData::from_urlencoded(payload);
        Ok(https_data.into())
    }

    /// Parse a SMS text. See [`SmsData::from_text`].
    pub fn from_text_sms<S: AsRef<str>>(text_sms: S) -> Result<Self, AmlError> {
        let sms_data = SmsData::from_text(text_sms)?;
        Ok(sms_data.into())
    }

    /// Parse a SMS data. See [`SmsData::from_data`].
    pub fn from_data_sms(data_sms: &[u8]) -> Result<Self, AmlError> {
        let sms_data = SmsData::from_data(data_sms)?;
        Ok(sms_data.into())
    }

    /// Parse a base64 encoded SMS data. See [`AmlData::from_data_sms`].
    pub fn from_base64_sms<S: AsRef<[u8]>>(base64_sms: S)-> Result<Self, AmlError> {
        match base64::decode(base64_sms) {
            Ok(bin_sms) => Self::from_data_sms(&bin_sms),
            Err(_) => Err(AmlError::InvalidBase64),
        }
    }
}

impl From<SmsData> for AmlData {
    fn from(sms: SmsData) -> Self {
        AmlData {
            version: sms.header,
            emergency_number: sms.emergency_number,
            beginning_of_call: sms.beginning_of_call,
            latitude: sms.latitude,
            longitude: sms.longitude,
            accuracy: sms.accuracy,
            time_of_positioning: sms.time_of_positioning,
            confidence: sms.level_of_confidence,
            altitude: sms.altitude,
            vertical_accuracy: sms.vertical_accuracy,
            positioning_method: sms.positioning_method,
            imsi: sms.imsi,
            imei: sms.imei,
            network_mcc: sms.network_mcc,
            network_mnc: sms.network_mnc,
            home_mcc: sms.home_mcc,
            home_mnc: sms.home_mnc,
            language: sms.language,
            transport: "sms".to_string(),
            ..Default::default()
        }
    }
}

impl From<HttpsData> for AmlData {
    fn from(https_data: HttpsData) -> Self {
        AmlData {
            version: https_data.v,
            emergency_number: https_data.emergency_number,
            source_of_activation: https_data.source,
            beginning_of_call: https_data.time,
            latitude: https_data.location_latitude,
            longitude: https_data.location_longitude,
            time_of_positioning: https_data.location_time,
            altitude: https_data.location_altitude,
            floor: https_data.location_floor,
            positioning_method: https_data.location_source,
            accuracy: https_data.location_accuracy,
            vertical_accuracy: https_data.location_vertical_accuracy,
            confidence: https_data.location_confidence,
            bearing: https_data.location_bearing,
            speed: https_data.location_speed,
            device_number: https_data.device_number,
            model: https_data.device_model,
            imsi: https_data.device_imsi,
            imei: https_data.device_imei,
            iccid: https_data.device_iccid,
            home_mcc: https_data.cell_home_mcc,
            home_mnc: https_data.cell_home_mnc,
            network_mcc: https_data.cell_network_mcc,
            network_mnc: https_data.cell_network_mnc,
            language: https_data.device_languages,
            transport: "https".to_string(),
            ..Default::default()
        }
    }
}



src/sms.rs
use std::collections::HashMap;
use chrono::{DateTime, LocalResult, NaiveDateTime, TimeZone, Utc};
use crate::{seconds_to_utc, valid_list, AmlError};

const DATETIME_FORMAT: &str = "%Y%m%d%H%M%S";

#[derive(Debug, Default)]
pub struct  SmsData {
    /// The header shall appear at the beginning of the SMS message.
    /// This is the version of AML.
    pub header: Option<String>,

    /// The emergency number dialed (i.e. 112, 911, ...).
    pub emergency_number: Option<String>,

    /// The beginning of the emergency call (UTC).
    pub beginning_of_call: Option<DateTime<Utc>>,

    /// The WGS84 latitude in degrees. Latitude is truncated to 5 decimal points.
    pub latitude: Option<f64>,

    /// The WGS84 longitude in degrees. Longitude is truncated to 5 decimal points.
    pub longitude: Option<f64>,

    /// Accuracy of location in meters. An accuracy of 0 represents unknown.
    pub accuracy: Option<f64>,

    /// The date and time that the handset determined the location area specified in UTC.
    /// This field may be ignored if location or beginning of call fields are valued to None.
    pub time_of_positioning: Option<DateTime<Utc>>,

    /// The Level of Confidence is a percentage probability that the mobile handset is within the area being communicated.
    pub level_of_confidence: Option<f64>,

    /// Vertical location in meters (truncated to 1 decimal point).
    /// This field may be ignored if location field is valued to None.
    pub altitude: Option<f64>,

    /// Vertical accuracy in meters (truncated to 1 decimal point).
    /// Accuracy of 0 represents unknown.
    /// This field may be ignored if location field is valued to None.
    pub vertical_accuracy: Option<f64>,

    /// The method used to determine the location area.
    /// One char string valued with `"W"` (wifi), `"C"` (cell), `"G"` (GNSS), `"F"` (fused) or `"U"` (unknown).
    /// This field may be ignored if location fields are valued to None.
    pub positioning_method: Option<String>,

    /// The SIM card identifier of the handset that has made the emergency call.
    pub imsi: Option<String>,

    /// The identifier of the handset that made the emergency call.
    pub imei: Option<String>,

    /// Mobile Country Code, used to determine the network country that the emergency call was made on.
    pub network_mcc: Option<String>,

    /// Mobile Network Code, used to determine the mobile network used to make the emergency call.
    pub network_mnc: Option<String>,

    /// Home Mobile Country Code.
    pub home_mcc: Option<String>,

    /// Home Mobile Network Code.
    pub home_mnc: Option<String>,

    /// Language tags (IETF BCP 47).
    pub language: Option<String>,

    /// (v1) The length of the entire SMS message including the header and the length attribute.
    pub message_length: Option<usize>,    

    /// SMS AML is validated for v1 if message length is equal to message_length.
    /// For v2, SMS AML is always validated. 
    pub is_validated: bool,
}

impl SmsData {
    /// Parse a SMS data.
    ///
    /// # Example
    ///
    /// ```
    /// use aml_lib::SmsData;
    /// 
    /// let input = "415193D98BEDD8F4DEECE6A2C962B7DA8E7DEEB56232990B86A3D9623B39B92783EDE86F784F068BD560B6D80C1683E568B81D7BDCB3E176F076EFB89BA77B39DCCD56A3C966B15D39DD9BD570B2590E56CBC168B21A4DB66B8FC7BD590CB66BBBC73D990DB66BB37B31D90C";
    /// let decoded = hex::decode(input).expect("Decoding failed");
    ///
    /// let sms_data = SmsData::from_data(&decoded);
    /// if let Ok(sms) = sms_data {
    ///     assert_eq!(sms.latitude, Some(37.42175));
    /// }
    /// ```
    pub fn from_data(bin_sms: &[u8]) -> Result<Self, AmlError> {
        let raw_sms: Vec<u8>;
        let text_sms: &str;

        raw_sms = Self::decode_7to8(bin_sms);
        text_sms = std::str::from_utf8(&raw_sms).unwrap_or_default();
        Self::from_text(text_sms)
    }

    /// Parse a SMS text.
    ///
    /// # Example
    ///
    /// ```
    /// use aml_lib::SmsData;
    /// 
    /// let sms_text = String::from(r#"A"ML=1;lt=48.82639;lg=-2.36619;rd=52;top=20191112112928;lc=68;pm=G;si=208201771948415;ei=353472104343540;mcc=208;mnc=20;ml=126"#);
    ///
    /// let sms_data = SmsData::from_text(&sms_text);
    /// if let Ok(sms) = sms_data {
    ///     assert_eq!(sms.latitude, Some(48.82639));
    /// }
    /// ```
    pub fn from_text<S: AsRef<str>>(text_sms: S) -> Result<Self, AmlError> {
        let properties = Self::get_properties(text_sms.as_ref());

        match properties.get(r#"A"ML"#) {
            Some(&"1") => {
                let mut sms_data = Self::from_text_v1(properties);
                if let Some(len) = sms_data.message_length {
                    sms_data.is_validated = len == text_sms.as_ref().len();
                };
                Ok(sms_data)
            },
            Some(&"2") => {
                let mut sms_data = Self::from_text_v2(properties);
                // By default AML SMS v2 is validate
                sms_data.is_validated = true;
                Ok(sms_data)
            },
            _ => Err(AmlError::UnimplementedVersion),
        }
    }

    fn from_text_v1(properties: HashMap<&str, &str>) -> Self {
        let mut sms: SmsData = Default::default();

        for (key, value) in properties {
            match (key, value) {
                (r#"A"ML"#, _) => sms.header = Some(value.to_string()),
                ("lg", _) => sms.longitude = value.parse::<f64>().ok(),
                ("lt", _) => sms.latitude = value.parse::<f64>().ok(),
                ("rd", _) => sms.accuracy = value.parse::<f64>().ok(),
                ("top", _) => {
                    if let Ok(ndt) = NaiveDateTime::parse_from_str(&value, DATETIME_FORMAT) {
                        sms.time_of_positioning = Some(DateTime::<Utc>::from_utc(ndt, Utc));
                    }
                }
                ("lc", _) => sms.level_of_confidence = value.parse::<f64>().ok(),
                ("pm", _) => {
                    sms.positioning_method =
                        valid_list!(value.to_uppercase(), "G", "W", "C", "U")
                }
                ("si", _) => sms.imsi = Some(value.to_string()),
                ("ei", _) => sms.imei = Some(value.to_string()),
                ("mcc", _) => sms.network_mcc = Some(value.to_string()),
                ("mnc", _) => sms.network_mnc = Some(value.to_string()),
                ("ml", _) => sms.message_length = value.parse::<usize>().ok(),
                (_, _) => (),
            }
        }

        sms
    }

    fn from_text_v2(properties: HashMap<&str, &str>) -> Self {
        let mut sms: SmsData = Default::default();
        let (mut et_opt, mut lt_opt): (Option<i64>, Option<i64>) = Default::default();

        for (key, value) in properties {
            match (key, value) {
                (r#"A"ML"#, _) => sms.header = Some(value.to_string()),
                ("en", _) => sms.emergency_number = Some(value.to_string()),
                ("et", _) => et_opt = value.parse::<i64>().ok(),
                ("lo", _) => {
                    let mut values: Vec<Option<f64>> =
                        value.split(',').map(|i| i.parse::<f64>().ok()).collect();
                    values.resize(3, None);
                    sms.latitude = values[0];
                    sms.longitude = values[1];
                    sms.accuracy = values[2];
                }
                ("lt", _) => lt_opt = value.parse::<i64>().ok(),
                ("lc", _) => sms.level_of_confidence = value.parse::<f64>().ok(),
                ("lz", _) => {
                    let mut values: Vec<Option<f64>> =
                        value.split(',').map(|i| i.parse::<f64>().ok()).collect();
                    values.resize(2, None);
                    sms.altitude = values[0];
                    sms.vertical_accuracy = values[1];
                }
                ("ls", _) => {
                    sms.positioning_method =
                        valid_list!(value.to_uppercase(), "G", "W", "C", "U", "F")
                }
                ("ei", _) => sms.imei = Some(value.to_string()),
                ("nc", _) => {
                    sms.network_mcc = value.get(..3).map(|s| s.to_string());
                    sms.network_mnc = value.get(3..).map(|s| s.to_string());
                }
                ("hc", _) => {
                    sms.home_mcc = value.get(..3).map(|s| s.to_string());
                    sms.home_mnc = value.get(3..).map(|s| s.to_string());
                }
                ("lg", _) => sms.language = Some(value.to_string()),
                (_, _) => (),
            }
        }

        if let Some(et) = et_opt {
            sms.beginning_of_call = seconds_to_utc!(et);
            if let Some(lt) = lt_opt {
                sms.time_of_positioning = seconds_to_utc!(et + lt);
            }
        }

        sms
    }

    fn get_properties(s: &str) -> HashMap<&str, &str> {
        s.split(';')
            .map(|property| {
                let key_value: Vec<&str> = property.split('=').collect();
                (key_value[0].trim(), key_value[1].trim())
            })
            .filter(|key_val| !key_val.0.is_empty() && !key_val.1.is_empty())
            .collect()
    }

    // The definition of the 7 bit encoding can be found in ETSI TS 123 038 (see clause 6.1.2.1.1 specifically)
    fn decode_7to8(raw_bytes: &[u8]) -> Vec<u8> {
        let (mut bits_len, mut bits) = (0_u8, 0_u8);
        let mut out = Vec::<u8>::with_capacity(raw_bytes.len() << 1);

        for byte in raw_bytes {
            out.push(((byte << bits_len) | bits) & 0x7F);
            bits = byte >> (7 - bits_len);
            bits_len += 1;

            if bits_len == 7 {
                out.push(bits);
                bits = 0;
                bits_len = 0;
            }
        }

        out
    }
}


src/hmac.rs
use sha1;

pub const SHA1_DIGEST_BYTES: usize = 20;
const SHA1_KEY_BYTES: usize = 64;

#[doc(hidden)]
pub fn hmac_sha1(key: &[u8], message: &[u8]) -> [u8; SHA1_DIGEST_BYTES] {
    // set constants for HMAC
    let inner_pad_byte: u8 = 0x36;
    let outer_pad_byte: u8 = 0x5c;
    let key_pad_byte:   u8 = 0x00;

    // instantiate internal structures
    let mut sha1_ctx = sha1::Sha1::new();
    let auth_key: &mut [u8; SHA1_KEY_BYTES] = &mut [key_pad_byte; SHA1_KEY_BYTES];

    // if the key is longer than the hasher's block length, it should be truncated using the hasher
    if key.len() > SHA1_KEY_BYTES {
        // derive new authentication from provided key
        sha1_ctx.update(key);

        // assign derived authentication key
        auth_key[..SHA1_DIGEST_BYTES].copy_from_slice(&(sha1_ctx.digest().bytes()));

        // reset hash for reuse
        sha1_ctx.reset();
    } else {
        auth_key[..key.len()].copy_from_slice(key);
    }

    // generate padding arrays
    let mut inner_padding: [u8; SHA1_KEY_BYTES] = [inner_pad_byte; SHA1_KEY_BYTES];
    let mut outer_padding: [u8; SHA1_KEY_BYTES] = [outer_pad_byte; SHA1_KEY_BYTES];

    for offset in 0..auth_key.len() {
        inner_padding[offset] ^= auth_key[offset];
        outer_padding[offset] ^= auth_key[offset];
    }

    // perform inner hash
    sha1_ctx.update(&inner_padding);
    sha1_ctx.update(message);
    let inner_hash = sha1_ctx.digest().bytes();
    sha1_ctx.reset();

    // perform outer hash
    sha1_ctx.update(&outer_padding);
    sha1_ctx.update(&inner_hash);
    sha1_ctx.digest().bytes()
}


src/https.rs
use std::borrow::Cow;
use chrono::{ DateTime, LocalResult, TimeZone, Utc };
use crate::{ millis_to_utc, valid_list, hmac::hmac_sha1 };

const HMAC_FIELD: &str = "hmac";

#[derive(Debug, Default)]
pub struct HttpsData {
    /// This is the version of AML.
    pub v: Option<String>,

    /// Emergency number dialed.
    pub emergency_number: Option<String>,

    /// Source of activation (call or sms).
    pub source: Option<String>,

    /// Version number for thunderbird module.
    pub thunderbird_version: Option<String>,

    /// Date and time of the beginning of call (UTC format).
    pub time: Option<DateTime<Utc>>,

    /// Ground truth latitude (for testing).
    pub gt_location_latitude: Option<f64>,

    /// Ground truth longitude* (for testing).
    pub gt_location_longitude: Option<f64>,

    /// The WGS84 latitude in degrees. Latitude is truncated to 5 decimal points.
    pub location_latitude: Option<f64>,

    /// The WGS84 longitude in degrees. Longitude is truncated to 5 decimal points.
    pub location_longitude: Option<f64>,

    /// The date and time that the handset determined the location area specified in UTC.
    pub location_time: Option<DateTime<Utc>>,

    /// Vertical location in meters.
    pub location_altitude: Option<f64>,

    /// Floor label (as in elevator button floor label - may be non-numeric).
    pub location_floor: Option<f64>,

    /// The method used to determine the location area. String valued with `wifi`, `cell`, `gps` or `unknown`.
    pub location_source: Option<String>,

    /// Location accuracy in meters.
    pub location_accuracy: Option<f64>,

    /// Vertical accuracy in meters.
    pub location_vertical_accuracy: Option<f64>,

    /// Confidence in location accuracy (percent between 0 and 1).
    pub location_confidence: Option<f64>,

    /// Bearing in degrees.
    pub location_bearing: Option<f64>,

    /// Speed in meters/second.
    pub location_speed: Option<f64>,

    /// Device phone number (may be mising).
    pub device_number: Option<String>,

    /// Device model.
    pub device_model: Option<String>,

    /// The SIM card identifier of the handset that has made the emergency call.
    pub device_imsi: Option<String>,

    /// The identifier of the handset that made the emergency call.
    pub device_imei: Option<String>,

    /// Integrated Circuit Card Identification number. 
    pub device_iccid: Option<String>,

    /// Home Mobile Country Code.
    pub cell_home_mcc: Option<String>,

    /// Home mobile Network Code.
    pub cell_home_mnc: Option<String>,

    /// Mobile Country Code, used to determine the network country that the emergency call was made on.
    pub cell_network_mcc: Option<String>,

    /// Mobile Network Code, used to determine the mobile network used to make the emergency call.
    pub cell_network_mnc: Option<String>,

    /// BCP 47 language tags (comma separated), in order from highest priority to lowest
    pub device_languages: Option<String>,

    /// Car crash date time
    adr_carcrash_time: Option<DateTime<Utc>>,

    /// Message Authentification Code
    pub hmac: Option<String>,
}

impl HttpsData {
    /// Verify the `hmac` field to authenticate the message.
    /// Assumes that HMAC is the last of the fields.
    ///
    /// ```
    /// use aml_lib::HttpsData;
    /// 
    /// const KEY: &str = "AML";
    ///
    /// let https = String::from(r#"v=1&device_number=%2B33611223344&location_latitude=0.85732&location_longitude=-4.26325&location_time=1604912121000&location_accuracy=10.4&location_source=GPS&location_certainty=83&hmac=f64c70eb238bb239e00e8ac8c023bf2b5d3c41dd"#);
    /// if HttpsData::is_authenticated(&https, KEY.as_bytes()) {
    ///     let data = HttpsData::from_urlencoded(&https);
    ///     /* Do something */
    /// }
    /// ```
    pub fn is_authenticated<S: AsRef<str>>(payload: S, key: &[u8]) -> bool {
        let splitted: Vec<&str> = payload.as_ref().split(&format!("&{}=", HMAC_FIELD)).collect();

        if splitted.len() != 2 { 
            return false; 
        }

        let message = splitted[0];
        let hmac = hex::encode(hmac_sha1(key, message.as_bytes()));

        hmac.eq(splitted[1])
    }

    /// Parse a HTTPS AML message. That assumes it is an URL encoded string. 
    ///
    /// ```
    /// use aml_lib::HttpsData;
    /// 
    /// let https = String::from(r#"v=1&device_number=%2B33611223344&location_latitude=0.85732&location_longitude=-4.26325&location_time=1604912121000&location_accuracy=10.4&location_source=GPS&location_certainty=83&hmac=f64c70eb238bb239e00e8ac8c023bf2b5d3c41dd"#);
    /// 
    /// let https_data = HttpsData::from_urlencoded(&https);
    /// assert_eq!(https_data.location_latitude, Some(0.85732));
    /// ```    
    pub fn from_urlencoded<S: AsRef<str>>(payload: S) -> Self {
        let mut https_data: HttpsData = Default::default();

        let attributes: Vec<(Cow<str>, Cow<str>)> =
            url::form_urlencoded::parse(payload.as_ref().as_bytes())
                .into_iter()
                .collect();

        for (key, value) in attributes {
            match (key.as_ref(), value.as_ref().trim()) {
                ("v", val) => https_data.v = Some(val.to_string()),
                ("emergency_number", val) => https_data.emergency_number = Some(val.to_string()),
                ("source", val) => {
                    https_data.source = valid_list!(val.to_lowercase(), "call", "sms")
                }
                ("thunderbird_version", val) => {
                    https_data.thunderbird_version = Some(val.to_string())
                }
                ("time", val) => https_data.time = millis_to_utc!(val),

                ("gt_location_latitude", val) => {
                    https_data.gt_location_latitude = val.parse::<f64>().ok()
                }
                ("gt_location_longitude", val) => {
                    https_data.gt_location_longitude = val.parse::<f64>().ok()
                }

                ("location_latitude", val) => {
                    https_data.location_latitude = val.parse::<f64>().ok()
                }
                ("location_longitude", val) => {
                    https_data.location_longitude = val.parse::<f64>().ok()
                }
                ("location_time", val) => https_data.location_time = millis_to_utc!(val),
                ("location_altitude", val) => {
                    https_data.location_altitude = val.parse::<f64>().ok()
                }
                ("location_source", val) => {
                    https_data.location_source =
                        valid_list!(val.to_lowercase(), "gps", "wifi", "cell", "unknown")
                }
                ("location_accuracy", val) => {
                    https_data.location_accuracy = val.parse::<f64>().ok()
                }
                ("location_vertical_accuracy", val) => {
                    https_data.location_vertical_accuracy = val.parse::<f64>().ok()
                }
                ("location_confidence", val) => {
                    https_data.location_confidence = val.parse::<f64>().ok()
                }
                ("location_bearing", val) => https_data.location_bearing = val.parse::<f64>().ok(),
                ("location_speed", val) => https_data.location_speed = val.parse::<f64>().ok(),

                ("device_number", val) => https_data.device_number = Some(val.to_string()),
                ("device_model", val) => https_data.device_model = Some(val.to_string()),
                ("device_imsi", val) => https_data.device_imsi = Some(val.to_string()),
                ("device_imei", val) => https_data.device_imei = Some(val.to_string()),
                ("device_iccid", val) => https_data.device_iccid = Some(val.to_string()),

                ("cell_home_mcc", val) => https_data.cell_home_mcc = Some(val.to_string()),
                ("cell_home_mnc", val) => https_data.cell_home_mnc = Some(val.to_string()),
                ("cell_network_mcc", val) => https_data.cell_network_mcc = Some(val.to_string()),
                ("cell_network_mnc", val) => https_data.cell_network_mnc = Some(val.to_string()),
                
                ("device_languages", val) => https_data.device_languages = Some(val.to_string()),
                ("adr_carcrash_time", val) => https_data.adr_carcrash_time = millis_to_utc!(val),
                ("hmac", val) => https_data.hmac = Some(val.to_string()),

                (_, _) => (),
            }
        }

        https_data
    }
}
